### PR TITLE
[DEVOPS-1029] Bump LTS-12

### PR DESCRIPTION
**Pull request by angerman**
## Description

Update the LTS version (and therefore the GHC version to 8.4). This is a prerequisite for cross compilation to windows.

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/DEVOPS-1029

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by angerman** [comment]
This contains the necessary changes to bump `develop` to lts-12.7 as a precursor for the windows cross compilation logic (which requires ghc8.4+, and as such lts-12+).

File: wallet-new/src/Cardano/Wallet/API/Request/Filter.hs [hunk 1]
```diff
@@ -349,6 +349,7 @@ instance
       where
         incorporate =
             foldr (uncurry appendToQueryString) req . toQueryString
+    hoistClientMonad _ _ _ _ = error "no hoistClientMonad here"
```
Comment (angerman) [inline wallet-new/src/Cardano/Wallet/API/Request/Filter.hs]: @parsonsmatt I'm pretty sure this is wrong. Do you happen to know what this should be?

haddocks at http://hackage.haskell.org/package/servant-client-core-0.14.1/docs/Servant-Client-Core-Internal-HasClient.html#v:hoistClientMonad

File: stack.yaml [hunk 2]
```diff
@@ -1,11 +1,13 @@
-resolver: lts-11.13
+resolver: lts-12.7
 
 flags:
   ether:
     disable-tup-instances: true
 
 extra-package-dbs: []
 
+allow-newer: true
```
Comment (angerman) [inline stack.yaml]: This needs to go.

@avieth I think you are the only one(?) who has access to https://github.com/input-output-hk/engine.io, could you give access to the `cardano-sl` group? The engine-io-wai module needs a version bump; and we might want to merge in the version bumps from upstream as well.

**Comment by angerman** [comment]
What?

**Comment by KtorZ** [comment]
Wut o.O ... 
Ho. I deleted `develop-new`. That might be it.

File: lib/src/Pos/Util/Servant.hs [hunk 3]
```diff
@@ -320,17 +320,20 @@ instance HasSwagger (subApi :> res) =>
 instance HasClient m (subApi :> res) => HasClient m (CDecodeApiArg subApi :> res) where
     type Client m (CDecodeApiArg subApi :> res) = Client m (subApi :> res)
     clientWithRoute p _ req = clientWithRoute p (Proxy @(subApi :> res)) req
+    hoistClientMonad _ _ _ _ = error "no hoistClientMonad here"
```
Comment (parsonsmatt) [inline lib/src/Pos/Util/Servant.hs]: man i don't really know what this one is doing, to be honest

Comment (parsonsmatt) [reply lib/src/Pos/Util/Servant.hs]: I would be tempted to just use the ones defined for `QueryParam` and friends:

```haskell
  hoistClientMonad pm _ f cl = \as ->
    hoistClientMonad pm (Proxy :: Proxy api) f (cl as)
```

File: wallet-new/src/Cardano/Wallet/API/Request/Filter.hs (hunk 1 above)
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/API/Request/Filter.hs]: as above, I'd use the ones form the query params:

```haskell
  hoistClientMonad pm _ f cl = \as ->
    hoistClientMonad pm (Proxy :: Proxy api) f (cl as)
```

File: wallet/src/Pos/Wallet/Web/Server/Handlers.hs [hunk 4]
```diff
@@ -14,9 +14,8 @@ import           Ntp.Client (NtpStatus)
 
 import           Pos.Wallet.Web.Swagger.Spec (swaggerSpecForWalletApi)
 import           Servant.API ((:<|>) ((:<|>)))
-import           Servant.Generic (AsServerT, GenericProduct, ToServant,
```
Comment (CodiePP) [inline wallet/src/Pos/Wallet/Web/Server/Handlers.hs]: compilation of this file really took a long time! (some minutes)
will this change bring this down to a few seconds?
would be interesting to time it..
(@parsonsmatt suggested that Servant.Generic is probably the critical piece)

Comment (angerman) [reply wallet/src/Pos/Wallet/Web/Server/Handlers.hs]: Myabe, maybe not. I didn't benchmark it. I've just tried hard to ensure it build with LTS-12

**Comment by angerman** [comment]
@erikd, @CodiePP, @parsonsmatt Please review one final time. :-)

File: nix/patches/ghc-8.0.2-darwin-rec-link.patch [hunk 5]
```diff
@@ -2,23 +2,12 @@ diff --git a/compiler/main/DriverPipeline.hs b/compiler/main/DriverPipeline.hs
 index acd0d61..3e83c15 100644
 --- a/compiler/main/DriverPipeline.hs
 +++ b/compiler/main/DriverPipeline.hs
-@@ -1916,6 +1916,7 @@ linkBinary' staticLink dflags o_files dep_packages = do
+@@ -1916,5 +1916,5 @@ linkBinary' staticLink dflags o_files dep_packages = do
                        ++ pkg_framework_opts
                        ++ debug_opts
                        ++ thread_opts
+-                      ++ (if (platformOS platform `elem` [OSDarwin]) then [ "-Wl,-dead_strip_dylibs" ] else [])
 +                      ++ (if (platformOS platform `elem` [OSDarwin, OSiOS]) then [ "-Wl,-dead_strip_dylibs", "-Wl,-dead_strip" ] else [])
```
Comment (rvl) [inline nix/patches/ghc-8.0.2-darwin-rec-link.patch]: Is this a patch for ghc 8.4.4? Can we rename the file?

File: stack.yaml [hunk 6]
```diff
@@ -60,8 +60,8 @@ packages:
   extra-dep: true
 
 - location:
-    git: https://github.com/input-output-hk/cardano-crypto
-    commit: e87fcd3fea18f664d6e4f39c11abb8cafbd25785
+    git: https://github.com/angerman/cardano-crypto
```
Comment (rvl) [inline stack.yaml]: Do you not have access to this repo?
Can we put a TODO in there to explain why the fork and reminder to move it back to input-output-hk.

File: .buildkite/pipeline.yml [hunk 7]
```diff
@@ -1,11 +1,11 @@
 steps:
   - label: 'stack2nix'
-    command: 'nix-shell -Q -j 4 --run scripts/check-stack2nix.sh'
+    command: 'nix-shell -Q -j 4 -E "with import (import ./nix/fetch-nixpkgs.nix) {}; runCommand  \"shell\" {} \"\"" --run scripts/check-stack2nix.sh'
```
Comment (rvl) [inline .buildkite/pipeline.yml]: I don't think this magic is needed any more.
```suggestion
    command: 'scripts/check-stack2nix.sh'
```

File: .buildkite/pipeline.yml [hunk 8]
```diff
@@ -1,11 +1,11 @@
...
     agents:
       system: x86_64-linux
 
   - label: 'release.nix'
-    command: 'nix-shell --run scripts/ci/check-hydra.sh'
+    command: 'nix-shell -E "with import (import ./nix/fetch-nixpkgs.nix) {}; runCommand  \"shell\" {} \"\"" --run scripts/ci/check-hydra.sh'
```
Comment (rvl) [inline .buildkite/pipeline.yml]: as above

File: stack.yaml [hunk 9]
```diff
@@ -81,7 +81,7 @@ packages:
 # https://github.com/well-typed/cborg/issues/171
 - location:
     git: https://github.com/well-typed/cborg
-    commit: 3d274c14ca3077c3a081ba7ad57c5182da65c8c1
+    commit: 37de2c0ae6bc7fa53f24f8d13748a7c370da9149
```
Comment (parsonsmatt) [inline stack.yaml]: Can you update the comment on this one indicating the PR that we're waiting on getting merged upstream?

File: stack.yaml [hunk 10]
```diff
@@ -170,7 +170,7 @@ packages:
 # buggy async
 - location:
     git: https://github.com/input-output-hk/log-warper.git
-    commit: 4af8114fbcd5698a4e1fbc653031f0300e29faee
+    commit: 0b7d4d48310f139d75829d31883aaa87ce53312c
```
Comment (parsonsmatt) [inline stack.yaml]: hmm, we're not using `log-warper` anymore I thought.

File: stack.yaml [hunk 11]
```diff
@@ -184,10 +184,10 @@ packages:
...
 # servant-quickcheck has a doc/doc.cabal file that is poorly specified and this
 # breaks the mafia build tool.
-- location:
-    git: https://github.com/avieth/servant-quickcheck.git
-    commit: e4e879abca99d0d56015765859fb37af59a81dbb
-  extra-dep: true
+# - location:
+#     git: https://github.com/avieth/servant-quickcheck.git
+#     commit: e4e879abca99d0d56015765859fb37af59a81dbb
+#   extra-dep: true
```
Comment (parsonsmatt) [inline stack.yaml]: Since this should be totally fixed up, we can delete the entry entirely.

File: stack.yaml [hunk 12]
```diff
@@ -214,90 +214,47 @@ nix:
...
-- diagrams-postscript-1.4.1
-- diagrams-svg-1.4.2
-- diagrams-solve-0.1.1
-- SVGFonts-1.6.0.3
-- dual-tree-0.2.1
-- fgl-5.5.4.0
-
-# Support for wallet's Txmeta tracking
-- sqlite-simple-errors-0.6.1.0
+#- Chart-diagrams-1.8.2
```
Comment (parsonsmatt) [inline stack.yaml]: Delete if it's not needed anymore :)
```suggestion
```

File: stack.yaml [hunk 13]
```diff
@@ -214,90 +214,47 @@ nix:
...
 
 # Working around a stack issue with rebuilding dependencies
-- pretty-show-1.7
+# - pretty-show-1.7
```
Comment (parsonsmatt) [inline stack.yaml]:
```suggestion
```

File: stack.yaml [hunk 14]
```diff
@@ -214,90 +214,47 @@ nix:
...
+- base58-bytestring-0.1.0
+- pipes-interleave-1.1.3
+- tabl-1.0.3
+- loc-0.1.3.3
+- hspec-2.5.5
+- micro-recursion-schemes-5.0.2.2
+- token-bucket-0.1.0.1
+
+# memory-0.14.18 exports reverse from Data.ByteArray
+# - memory-0.14.17
```
Comment (parsonsmatt) [inline stack.yaml]:
```suggestion
```

File: stack.yaml (hunk 10 above)
Comment (angerman) [reply stack.yaml]: I'll leave https://github.com/input-output-hk/cardano-crypto/pull/46 just here. But add it to the stack.yaml as well.

File: stack.yaml (hunk 6 above)
Comment (angerman) [reply stack.yaml]: I'll leave https://github.com/input-output-hk/cardano-crypto/pull/46 just here. But add it to the stack.yaml as well.

File: stack.yaml (hunk 9 above)
Comment (angerman) [reply stack.yaml]: Looks like this has been merged/released already. Let's try.

File: nix/patches/ghc-8.0.2-darwin-rec-link.patch (hunk 5 above)
Comment (angerman) [reply nix/patches/ghc-8.0.2-darwin-rec-link.patch]: this is not really 8.4.4 specific. It's just a more comprehensive patch than what nixpkgs upstream does.

File: stack.yaml (hunk 10 above)
Comment (angerman) [reply stack.yaml]: Still used by
- cardano-report-server-0.5.10 (due to cardano-sl-core-2.0.0)
- cardano-sl-faucet-2.0.0
- cardano-sl-utxo

Comment (erikd) [reply stack.yaml]: Those should probably be fixed to use the new logging method.

Don't worry about it for this PR. Maybe someone on @CodiePP 's team can take care of it.

File: stack.yaml (hunk 6 above)
Comment (rvl) [reply stack.yaml]: Cool, thanks.

File: stack.yaml (hunk 10 above)
Comment (CodiePP) [reply stack.yaml]: yes, good point. we will investigate ..

**Comment by rvl** [comment]
AppVeyor build is currently failing due to `MAX_PATH` ... oh woe! :cry:

**Comment by angerman** [comment]
Finally... using a patched compiler for windows, and building on our windows build kite slave instead of appveyor!

**Comment by rvl** [comment]
AppVeyor build works and the executables run correctly in Windows within Daedalus.

**Pull request merged**